pkg/apis/istio/v1alpha2/validation/validation_test.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validation
+
+import (
+	"reflect"
+	"testing"
+
+	"istio.io/operator/pkg/apis/istio/v1alpha2"
+)
+
+func makeBoolPtr(v bool) *bool {
+	return &v
+}
+func makeStringPtr(v string) *string {
+	return &v
+}
+
+func TestValidate(t *testing.T) {
+	tests := []struct {
+		name       string
+		toValidate *v1alpha2.Values
+		validated  bool
+	}{
+		{
+			name:       "Empty struct",
+			toValidate: &v1alpha2.Values{},
+			validated:  true,
+		},
+		{
+			name: "With CNI defined",
+			toValidate: &v1alpha2.Values{
+				CNI: &v1alpha2.CNIConfig{
+					Enabled: makeBoolPtr(true),
+				},
+			},
+			validated: true,
+		},
+		{
+			name: "With Slice",
+			toValidate: &v1alpha2.Values{
+				Gateways: &v1alpha2.GatewaysConfig{
+					Enabled: makeBoolPtr(true),
+					EgressGateway: &v1alpha2.EgressGatewayConfig{
+						Ports: []*v1alpha2.PortsConfig{
+							{
+								Name: makeStringPtr("port1"),
+							},
+							{
+								Name: makeStringPtr("port2"),
+							},
+						},
+					},
+				},
+			},
+			validated: true,
```

Comment:
test for validation failures?

Reply:
there are two parts of this test, 1st is the actual tree traversal and dealing with different types of vars without panic. 2nd is to test actual type validation. Right now all type validations are returning empty error slice, since there is no validation logic. Once actual type validation gets added then a failure scenario could be introduced by the test developer. The negative test will also test indirectly, whether type's validation was called or not. I understand that it is not ideal, but so fat I do not see how else type's validation can notify the test that it was called without either modification of type's validate func or making a copy of v1alpha2.Values in test. Both option do not look good to me, if you could suggest some alternative, I would be happy to use it.